### PR TITLE
add support for storyshots to save files according to different patterns

### PR DESCRIPTION
**Pull request by zvictor**
# The issue

In my setup, I always have a `cases.spec.js` that contains different props for my components. This file provides cases scenarios for both unit tests and storybook. With that in place, my components always have a standard `stories.js` file next to it, that looks like that:

```js
import React from 'react';
import { storiesOf } from '@storybook/react';

import Component from './component';
import cases from './cases.spec.js';

const stories = storiesOf('Category Icon Hover Menu', module)

for (const [caseName, caseData] of cases) {
  stories.addCentered(caseName, () => (
    <Component {...caseData} />
  ));
}
```

I like organising the project like this because I get storybook "for free" while writing tests. However, the `multiSnapshotWithOptions` does not work the way I intended with such setup. I don't like that it always create a `__snapshots__` folder with a **single file** on it.

screen shot 2017-11-30 at 13 25 24

# The alternative

I rewrote `getStoryshotFileName` to use the stories `kind` and `name` instead of the original filename:

screen shot 2017-11-30 at 13 31 20

The API is quite simple:

```js
import initStoryshots, { snapshotPerStoryAdded, snapshotPerStoryFile  } from '@storybook/addon-storyshots';

initStoryshots({
  // filesPattern: snapshotPerStoryAdded,
  filesPattern: snapshotPerStoryFile,
});
```

with that I also killed the `multiSnapshotWithOptions`. I hope nobody minds 😄

### CODE REVIEW

File: addons/storyshots/src/utils.js [hunk 1]
```diff
@@ -1,25 +1,47 @@
...
+    const { fileName, kind, story } = context;
+
+    if (!fileName) {
+      return null;
+    }
 
-  return getStoryshotFile(fileName);
-}
+    const { dir } = path.parse(fileName);
+    const name = sanitize(`${kind}--${story}`)
```
Comment (zvictor) [inline addons/storyshots/src/utils.js]: is `--` the separator we want to use?

**Review by ndelangen — APPROVED** [review]

**Comment by igor-dv** [comment]
I will test it soon too. Let's not merge for now. We want to fix things with angular in 3.3 before + we have [this](https://github.com/storybooks/storybook/pull/2413) that might conflict.

**Comment by zvictor** [comment]
> I will test it soon too. Let's not merge for now.

Cool. Just to make it clear, I am starting my vacations now and I will be out for a whole month. If you need me to make changes in the code it will need to wait. Sorry for that.

**Comment by igor-dv** [comment]
If it's not impeding your work - it's ok. I think there won't be a much work here during the holidays.

**Comment by ndelangen** [comment]
Hey @zvictor can you please fix the merge conflict?

**Review by igor-dv — CHANGES_REQUESTED** [review]
You can't really remove the 'multiSnapshotWithOptions' it's already in 3.3 release, so renaming/removing it is a breaking change.

**Comment by igor-dv** [comment]
@zvictor , when you are back, let's revisit the implementation. I find it very useful to customize file names and extensions.

**Comment by zvictor** [comment]
@igor-dv sorry for my absence. I have been more busy than I expected since I came back from vacation and to be honest I cannot estimate when I would have some time available. Nonetheless, some of the changes that were made in the project since the creation of this PR are beyond my comprehension, so it would be a bit hard for me to do a proper merge.

with all of that said, I would be more than happy if you could take over this PR and work on the needed changes.

**Comment by igor-dv** [comment]
Sure

**Comment by bialesdaniel** [comment]
Any updates on this? My use case is that I have multiple `index.stories.js` files and if one of the stories change it gives me multiple failed tests. Being able to name the file based on the parent directory or something would be helpful.

**Comment by igor-dv** [comment]
@bialesdaniel, I am working on it.
I am closing this PR since there are too many changes in the repo to make it work.

**Pull request closed without merging**